test/socket.test.js

```diff
@@ -0,0 +1,35 @@
+const { app } = require('../src/expressServer');
+const io = require('socket.io-client');
+
+let server;
+const PORT = 8001;
+
+beforeAll(async () => {
+  server = await app.listen(PORT);
+});
+
+afterAll(async () => {
+  server.close();
+});
+
+test('test:echo socket event', (done) => {
+  jest.setTimeout(30000);
+  const client = io.connect(`http://localhost:${PORT}`, {
+    transports: ['websocket'],
+    forceNew: true,
+    reconnection: false,
+  });
+
+  const messageToSend = 'Hello, server!';
+
+  client.once('connect', () => {
+    client.emit('test:echo', messageToSend);
+
+    client.on('test:echo', (response) => {
+      expect(response).toBe(messageToSend);
```

Comment:
If we create a `test:echo` event handler we will need to have it do `this.emit(<same message that was sent in>)`. `.emit` will send that message to _all_ sockets listening to that topic. 

But out current event handler for `test:ping` does something a little different: [https://github.com/Code-the-Dream-School/dd-prac-team4-back/blob/PS192-socket-unit-test/src/live/testHandlers.js#L5.](https://github.com/Code-the-Dream-School/dd-prac-team4-back/blob/PS192-socket-unit-test/src/live/testHandlers.js#L5.) It does `this.broadcast.emit("user sent: <message that was sent in>")` so not only is the string that's being emmited is different than what was sent in, it also uses `.broadcast.emit` which is a little different from `.emit`. With `.broadcast.emit` the message is sent to _all listeners EXCEPT the one who sent the message_. This can be useful for example with chat; the person who joined the chat doesn't need to get a notification that they themselves joined the chat; that message is only useful for _everyone else in the room_.

But in that case, to test it we need to have a second client to receive the message. We can create that second client in exactly the same way as lines 17-21 here. And then we can have something like:

```javascript
  senderClient.once('connect', () => {
    senderClient.emit('test:ping', ...);
  });

  recipientClient.once('connect', () => {
    recipientClient.on('test:ping', (response) => {
      expect(response).toBe(...);

      senderClient.disconnect();
      recipientClient.disconnect();

      // small timeout before calling "done()" to give time for the disconnect event to fire
      setTimeout(() => done(), 1000)
    });
  })
```

Reply:
Thank you very much, work now and I understand.  But I cant understand how to stop setInterval from Order models. I tried add clearInterval(intervalId) to afterAll, but not work.
<img width="661" alt="screen" src="https://github.com/Code-the-Dream-School/dd-prac-team4-back/assets/95896159/25d42b43-d076-4a40-8960-45ae339fa693">

Reply:
Thank you very much, work now and I understand.❤️